src/Runtime/OMUnique.inc

```diff
@@ -0,0 +1,422 @@
+              int64_t tdim[6] = {
+                  dim[0], dim[1], dim[2], dim[3], dim[4], dim[5]};
+              tdim[sliceAxis] = indicesPtr[dim[sliceAxis]];
+              // calculate the input offset and pointer
+              uint64_t inputOff = tdim[0] * strides[0] + tdim[1] * strides[1] +
+                                  tdim[2] * strides[2] + tdim[3] * strides[3] +
+                                  tdim[4] * strides[4] + tdim[5] * strides[5];
+              void *inputNumPtr = ((char *)inputPtr) + inputOff * dataSize;
+              // calculate the output offset and pointer
+              uint64_t outputOff =
+                  dim[0] * outStrides[0] + dim[1] * outStrides[1] +
+                  dim[2] * outStrides[2] + dim[3] * outStrides[3] +
+                  dim[4] * outStrides[4] + dim[5] * outStrides[5];
+              void *YNumPtr = ((char *)YPtr) + outputOff * dataSize;
+              memcpy(YNumPtr, inputNumPtr, dataSize);
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+//
+// "omTensorUnique" handles two cases, where are (1) case with no axis given
+// (2) case with axis given.
+// -  case with no axis given
+// In this case, omTensorUnique gets one element in the input Tensor in turn,
+// calls sliceTableRegister using an element as a slice. The registered
+// elements are used as the output Y.
+// - case with axis given
+// In this case, omTensorUnique gets a slice with getSliceData, registers it
+// with getSliceData, The outputs indices, inverse_indices and counts are set
+// by getSliceData. If the output Y is necessary, it is created by produceY.
+//
+void omTensorUnique(OMTensor *totalTensor, OMTensor *Y, OMTensor *indices,
+    OMTensor *inverse_indices, OMTensor *counts, const OMTensor *inputTensor,
+    int64_t sliceAxis, uint64_t sorted) {
+  const OM_DATA_TYPE dataType = omTensorGetDataType(inputTensor);
+  const int64_t inputRank = omTensorGetRank(inputTensor);
+  assert(inputRank <= 6 && "input rank should be 6 or less");
+  assert(sliceAxis < inputRank && "axis should be less than rank");
+  int64_t *totalPtr = (int64_t *)omTensorGetDataPtr(totalTensor);
+  const int64_t *inputShape = omTensorGetShape(inputTensor);
+  void *inputPtr = omTensorGetDataPtr(inputTensor);
+  void *YPtr = (Y != NULL) ? omTensorGetDataPtr(Y) : NULL;
+  void *indicesPtr = (indices != NULL) ? omTensorGetDataPtr(indices) : NULL;
+  void *inverseIndicesPtr =
+      (inverse_indices != NULL) ? omTensorGetDataPtr(inverse_indices) : NULL;
+  void *countsPtr = (counts != NULL) ? omTensorGetDataPtr(counts) : NULL;
+  uint64_t dataSize = OM_DATA_TYPE_SIZE[dataType];
+
+  int count = 0;
+  sliceTable sliceTable;
+  if (sliceAxis < 0) { // if slice attribute is not specified
+    // manage the inputTensor as flatten one
+    uint64_t elementNum = 1;
+    for (int64_t i = 0; i < inputRank; i++)
+      elementNum *= inputShape[i];
+    void *sliceDataPtr = (YPtr == NULL) ? alloca(dataSize * elementNum) : YPtr;
```

Comment:
it seems unsafe to use alloca() for things of unbounded size

would it be better to change this to malloc()?

Reply:
@negiyas please address @sorenlassen point. For large data, `alloc` is safer. Please open a new PR for this. Thanks